### ISSUE TITLE
Exclude / Ignore changes

### ISSUE THREAD

**Issue by funwithtriangles** [charkour/zundo#8]
It would be great to have the ability to ignore certain top level properties in the state tree, so they are never affected by undo and redo actions. Currently this is achievable with multiple stores but I'm not too fond of this solution.

I realise this has already been raised (https://github.com/charkour/zundo/issues/5) but that's a mixed issue so I figured it made more sense to have a standalone issue for this subject.

**Comment by charkour** [comment]
Hi! Thanks fir opening the issue. Yeah, #5 is a combined issue so thanks for opening this one. 

I agree it would be a nice feature to have. What would you like the API to look like?

Also, out of curiosity, what is the issue of making multiple stores? One for undoable data and the other for not? 

Thanks

**Comment by funwithtriangles** [comment]
Hey! Thanks for getting back :) I think passing an array of property names for the top-level parts of the state you want to ignore is quite a simple approach. I was also thinking you could modify the `set` function to have an extra argument where you can override whether that particular change to state gets stored in history or not. If the user hits "back", it would simply skip over that point in time. This would allow for more fine grained control and is probably a bit more of an advanced use case, but could be quite nice. 

What do you think the best approach would be? I think both of those methods have different use cases and don't actually clash with each other, so you could even have both.

The main reason for not wanting to have multiple stores is that I don't want to have to think about whether I need to use `useUndoableStore` and `useIgnoredStore` in my components, I don't feel like that kind of logic should be kept there. Here's a few other reasons I can think of:

- Generally having to change the structure of your state management just for this feature feels weird to me, middlewares are there to handle that without much pain. It's nice to be able to change the way undo/redo behaves without having to move parts of state to different stores
- I might actually need that advanced API I've mentioned where I need to ignore some deep part of state which can't be extracted easily into a separate store (e.g. have a specialised `set` function where certain changes to state can be completely removed from history)
- It's nicer with Redux devtools to see everything in one place (I think it does support multiple stores but across tabs or something weird like that)

**Comment by charkour** [comment]
Great, thanks for the detailed response! 

Alright, so for omitting fields from being tracked, something like this would make sense?

```tsx 

import create, { UndoState } from 'zundo';

// creates a store with undo/redo capability
const useStoreWithUndo = create<StoreState>(set => ({
  // other fields defined
  actionFunction: () => set(state => {...state}, ["fieldName", "fieldName2"])
}));

```

This API has modified the `set` function to take a second parameter which is an array of field names to omitting from tracking history. This is what you're suggesting? I think that would be great to add. 

Thanks for explaining why it makes sense to not create multiple stores. I agree with you. 

Would you please make an issue to track the other feature request for modifying `set` to conditionally store previous states. I think I understand your request and it would be good to have; although I want to talk through it more to make sure we are on the same page. 

Thanks!

**Comment by funwithtriangles** [comment]
Ah, I was actually thinking that array gets passed in on store creation. I
can make another issue explaining the modification of set.

**Comment by charkour** [comment]
Thanks. 

Something like this?

```tsx 

import create from 'zundo';

// creates a store with undo/redo capability
const useStoreWithUndo = create(set => ({
  // other fields defined
  actionFunction: () => set(state => {...state})
}), ["fieldName", "fieldName2"]);

```

**Comment by funwithtriangles** [comment]
Yep! What do you reckon? I think this makes sense for the basic ignoring of
state slices.

**Comment by charkour** [comment]
Yeah, let's go with that. I'll spend some time tomorrow afternoon implementing it. Thanks

**Comment by charkour** [comment]
I worked on it. Testing it now and will have a PR in the next few days.

**Comment by charkour** [comment]
Hey @funwithtriangles, PR #10 implement your request. Would you be willing to check it out to see if it works the way you were hoping?

If you have any ideas on how you could test my changes more easily, please let me know.

Thanks!

**Comment by funwithtriangles** [comment]
Hey! I'm working on a different project today, will pick this up on Monday :)

**Comment by funwithtriangles** [comment]
Tested in my project and it's all working great, even with other middlewares :)

**Comment by charkour** [comment]
Thanks

**Comment by charkour** [comment]
This is released in v1.1.0
